modules/registration/proxy_request.rb

```diff
@@ -10,11 +10,24 @@ def global_register(request)
       send_request(proxy_req)
     end
 
+    # we support two way of sending data - either a JSON or url encoded data
     def host_register(request)
-      proxy_req = request_factory.create_post '/register',
-                                              request.body.read,
-                                              headers(request),
-                                              request_params(request)
+      if request.content_type == 'application/x-www-form-urlencoded'
+        # the request has a different content type, ForemanRequestProxy sets content type to json, unless
+        # specified explicitly
+        # also request.params contain the same data that is in request.body, just parsed to hash, so we need
+        # to remove them
+        proxy_req = request_factory.create_post '/register',
+                                                request.body.read,
+                                                headers(request).merge("Content-Type" => "application/x-www-form-urlencoded"),
```

Comment:
This looks odd - it implies we filter it out which shouldn't happen, right?

Reply:
I'm not sure what you mean, but I'll try to explain a bit better

1) headers don't contain "Content-Type", this is how we build the headers list https://github.com/theforeman/smart-proxy/blob/develop/modules/registration/proxy_request.rb#L30-L35, we could keep it there, however that would always override the default that we later set (see point 2)
2) the problem then is https://github.com/theforeman/smart-proxy/blob/develop/lib/proxy/request.rb#L29, which defaults to JSON if no content-type is specified in headers, normally you set the content-type directly on request object

Therefore we need to add this explicitly.

Reply:
> headers don't contain "Content-Type", this is how we build the headers list https://github.com/theforeman/smart-proxy/blob/develop/modules/registration/proxy_request.rb#L30-L35, we could keep it there, however that would always override the default that we later set (see point 2)

So this is the confusing part for me: I assumed that the header was present. The problem here is normalization.

The request comes in with the env var `HTTP_CONTENT_TYPE`, which remains. So `header_request` will have `CONTENT_TYPE` set, but in point 2 as you pointed out only looks at `Content-Type`. It's not case insensitive (as HTTP headers are) and only looks at `-` instead of `_`. At least, that's my theory but I didn't check it out.

Reply:
ah it seems this will not work reliably though, the `add_headers` method adds the content type twice then which causes tests to fail, I'll need to find a better way how to set the custom content_type, probably the change won't be just in the registration module :-(

Reply:
> The request comes in with the env var HTTP_CONTENT_TYPE, which remains.

it does not, this is the result form pry above this line
```ruby
{"USER_AGENT"=>"curl/7.69.1",
 "ACCEPT"=>"*/*",
 "AUTHORIZATION"=>
  "Bearer eyJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjo0LCJpYXQiOjE2MDM3NDAxMDcsImp0aSI6ImUzYThjZjZiYTY0MzgxZGE4ODAwZTEzZWY0NGYwMGVlMmZiMGU5NDZhYzExZjRiYTFmYjFmMTc5ZDA5NjczNzgiLCJleHAiOjE2MDM3NTQ1MDd9.DgnqIoWrhSZYsHqeUxY-08jwLtB5iBD8rApUrUuVYuM"}
```

request.env only contains these that start with HTTP
```ruby
{"HTTP_HOST"=>"foreman.example.tst:8443",
 "HTTP_USER_AGENT"=>"curl/7.69.1",
 "HTTP_ACCEPT"=>"*/*",
 "HTTP_AUTHORIZATION"=>
  "Bearer eyJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjo0LCJpYXQiOjE2MDM3NDAxMDcsImp0aSI6ImUzYThjZjZiYTY0MzgxZGE4ODAwZTEzZWY0NGYwMGVlMmZiMGU5NDZhYzExZjRiYTFmYjFmMTc5ZDA5NjczNzgiLCJleHAiOjE2MDM3NTQ1MDd9.DgnqIoWrhSZYsHqeUxY-08jwLtB5iBD8rApUrUuVYuM",
 "HTTP_VERSION"=>"HTTP/1.1"}
```

`CONTENT_TYPE` is not prefixed with HTTP. But you're right, even if we kept that header, it wouldn't match because of case sensitivity.

There's another problem with `add_headers` method I mentioned in previous post, we must avoid adding the content-type twice. New version of the patch contains the fix for it. I also updated the git commit message for this change.